development/gcp/cloud-functions/rotateserviceaccount/rotateserviceaccount.go

```diff
@@ -0,0 +1,125 @@
+	if err != nil {
+		panic("failed creating IAM client, error: " + err.Error())
+	}
+}
+
+// RotateServiceAccount adds new secret version in Secret Manager on pubsub rotate message
+func RotateServiceAccount(ctx context.Context, m pubsub.MessagePayload) error {
+	var err error
+	var secretRotateMessage pubsub.SecretRotateMessage
+	var secretData ServiceAccountJSON
+
+	// Create logger to use google cloud functions structured logging
+	logger := cloudfunctions.NewLogger()
+	// Set component for log entries to identify all messages for this function.
+	logger.WithComponent("kyma.prow.cloud-function.RotateServiceAccount")
+	// Set trace value for log entries to identify messages from one function call.
+	logger.GenerateTraceValue(projectID, "RotateServiceAccount")
+
+	if m.Attributes["eventType"] != "SECRET_ROTATE" {
+		logger.LogDebug("Unsupported event type: %s, quitting", m.Attributes["eventType"])
+		return nil
+	}
+
+	err = json.Unmarshal(m.Data, &secretRotateMessage)
+	if err != nil {
+		logger.LogCritical("failed to unmarshal message data field, error: %s", err.Error())
+	}
+
+	if secretRotateMessage.Labels["type"] != "service-account" {
+		logger.LogDebug("Unsupported secret type: %s, quitting", secretRotateMessage.Labels["type"])
+		return nil
+	}
+
+	//get latest secret version data
+	secretlatestVersionPath := secretRotateMessage.Name + "/versions/latest"
+	logger.LogInfo("Retrieving secret: %s", secretlatestVersionPath)
+	secretDataString, err := secretManagerService.GetSecretVersionData(secretlatestVersionPath)
+	if err != nil {
+		logger.LogCritical("failed to retrieve latest version of a secret %s, error: %s", secretRotateMessage.Name, err.Error())
+	}
+
+	logger.LogInfo("Trying to unmarshal secret: %s", secretRotateMessage.Name)
+	decodedSecretDataString, err := base64.StdEncoding.DecodeString(secretDataString)
+	if err != nil {
+		logger.LogCritical("Could not base64 decode secret %s", secretRotateMessage.Name)
+	}
+	err = json.Unmarshal([]byte(decodedSecretDataString), &secretData)
+	if err != nil {
+		logger.LogCritical("failed to unmarshal secret JSON field, error: %s", err.Error())
+	}
+
+	// get client_email
+	serviceAccountPath := "projects/" + secretData.ProjectID + "/serviceAccounts/" + secretData.ClientEmail
+	logger.LogInfo("Looking for service account %s", serviceAccountPath)
+	createKeyRequest := iam.CreateServiceAccountKeyRequest{}
+	newKeyCall := serviceAccountService.Projects.ServiceAccounts.Keys.Create(serviceAccountPath, &createKeyRequest)
+	newKey, err := newKeyCall.Do()
+	if err != nil {
+		logger.LogCritical("failed to create new key for %s Service Account, error: %s", serviceAccountPath, err.Error())
+	}
+
+	logger.LogInfo("Decoding new key data for %s", serviceAccountPath)
+	newKeyBytes, err := base64.StdEncoding.DecodeString(newKey.PrivateKeyData)
+	if err != nil {
+		logger.LogCritical("failed to decode new key for %s Service Account, error: %s", serviceAccountPath, err.Error())
+	}
```

Comment:
Same here. Creating and decoding a new key will be frequent usage of serviceAccountService. You can make it a method of serviceAccountService and call it here with respective args.

Reply:
I don't think this will be a frequent usage, as outside of this function we shouldn't need to create or rotate service account keys